Add network storage documentation (#60)
symfonycorp/symfony-cloud-docs: services/network-storage.rst

Network Storage
===============

The network storage service provides a file store and enables a new kind of
``mount`` that can be shared between different application containers.
application containers.

To use it in your application, add it to ``.symfony/services.yaml``:

.. code-block:: yaml

    files:
        type: network-storage:1.0
        disk: 256

And wire it in ``.symfony.cloud.yaml``:

.. code-block:: yaml

    mounts:

>         "/var": { source: local, source_path: var }

Reviewer: ```suggestion
        '/var': { source: local, source_path: var }
```

Author: @fabpot you may want to update this in the fast track as well

>         "/public/uploads": { source: service, service: files, source_path: uploads }

Reviewer: ```suggestion
        '/public/uploads': { source: service, service: files, source_path: uploads }
```

.. note::

   You do **not** need to add a relationship to point to the ``files`` service.
   This is handled automatically by SymfonyCloud.

The application can now read from and write to the ``public/uploads`` directory
as if it was a local writeable mount.

.. note::

>     There is a small performance hit for using a network mount over a local

Reviewer: ```suggestion
    There is a small performance impact for using a network mount over a local
```

>     mount but in most cases it should not be noticeable. Please note that

Reviewer: ```suggestion
    mount, but in most cases it should not be noticeable. Please note that
```

    high-volume sequential file creation may experience a more significant

>     performance hit.

Reviewer: ```suggestion
    performance impact.
```

> How can I migrate a local storage to a network storage?

Reviewer: ```suggestion
How can I migrate a Local Storage to a Network Storage?
```

-------------------------------------------------------

There is no automated way of transferring data from one storage type to
another. Fundamentally, the process is consists in moving files around.

Suppose you have this mount configuration and want to move to a network storage
mount:

.. code-block:: yaml

    mounts:

>         "/var": { source: local, source_path: var }

Reviewer: ```suggestion
        '/var': { source: local, source_path: var }
```

>         "/public/uploads": { source: local, source_path: uploads }

Reviewer: ```suggestion
        '/public/uploads': { source: local, source_path: uploads }
```

The following approximate steps will do so with a minimum of service interruption.

> 1. Add a new ``network-storage`` service named `files`, that has at least

Reviewer: ```suggestion
1. Add a new ``network-storage`` service named ``files``, that has at least
```

>    enough space for your existing files with some buffer

Reviewer: ```suggestion
   enough space for your existing files with some buffer.
```

   .. note::

>        You may need to increase your project's allocated storage to do

Reviewer: ```suggestion
       You may need to increase your project's allocated storage (at least for the transition phase) to do
```

Author: increate project's storage allocation can not be reverted, I believe this addition could be misleading and make users think it could be reverted.

Reviewer: Oh ok then it should be a .. caution to make this clear

Author: makes sense

Author: updated

>        this.

Reviewer: ```suggestion
       this, at least for the transition phase.
```

2. Add a new mount to the network storage service on a non-public directory:

>    .. code-block:: yaml

Reviewer: Maybe it makes sense here to use the `diff` code-block like:

```rst
    .. code-block:: diff

        mounts:
        "/var": { source: local, source_path: var }
        "/public/uploads": { source: local, source_path: uploads }
+        "/new-uploads": { source: service, service: files, source_path: uploads }
```

Author: indeed 👍

       mounts:

>            "/var": { source: local, source_path: var }

Reviewer: ```suggestion
           '/var': { source: local, source_path: var }
```

>            "/public/uploads": { source: local, source_path: uploads }

Reviewer: ```suggestion
           '/public/uploads': { source: local, source_path: uploads }
```

>            "/new-uploads": { source: service, service: files, source_path: uploads }

Reviewer: ```suggestion
           '/new-uploads': { source: service, service: files, source_path: uploads }
```

> 3. Deploy the changes.

Reviewer: ```suggestion
3. Deploy the changes;
```

> 4. Connect to your container using ``symfony ssh``

Reviewer: ```suggestion
4. Connect to your container using ``symfony ssh``;
```

5. Use ``rsync`` to copy all files from the local mount to the network mount:

   .. code-block:: terminal

>       $ rsync -avz web/uploads/* new-uploads/

Reviewer: No ``symfony``-prefix here? 🤔

> 5. Reverse the mounts, commit and deploy:

Reviewer: ```suggestion
6. Reverse the mounts, commit and deploy:
```

>    .. code-block:: yaml

Reviewer: Maybe use the `diff` code-block again?

       mounts:
           "/var": { source: local, source_path: var }
           "/old-uploads": { source: local, source_path: uploads }
           "/public/uploads": { source: service, service: files, source_path: uploads }

> 5) Run rsync one more time to make sure any files uploaded during the

Reviewer: ```suggestion
6. Run ``rsync`` again to make sure all files uploaded during the transition are moved.
```

>    transition are not lost.

Reviewer: ```suggestion

```

   .. code-block:: terminal

      $ rsync -avz old-uploads/* web/uploads/

> 6. Once you're confident all the files are accounted for, delete the entire

Reviewer: ```suggestion
8. Once you're confident all the files are accounted for, delete the entire
```

>    contents of ``old-uploads``. If you do not, the files will remain on disk

Reviewer: ```suggestion
   contents of ``old-uploads``. If you do not, the files will remain unaccessible on disk
```

Author: we lose the idea that data would use storage space.

>    but inaccessible, just eating up disk space needlessly.

Reviewer: ```suggestion

```

> 7. Once that is done you can remove the ``old-uploads`` mount and push again to

Reviewer: ```suggestion
9. Once that is done you can remove the ``old-uploads`` mount and push again to
```

Author: good catch :)

Network Storage
===============

The network storage service provides a file store and enables a new kind of
``mount`` that can be shared between different application containers.

> application containers.

Reviewer 2: Looks like this should be removed

To use it in your application, add it to ``.symfony/services.yaml``:

.. code-block:: yaml

    files:
        type: network-storage:1.0
        disk: 256

And wire it in ``.symfony.cloud.yaml``:

.. code-block:: yaml

    mounts:
        '/var': { source: local, source_path: var }
        '/public/uploads': { source: service, service: files, source_path: uploads }

> .. note::

Reviewer 2: I would move that note after the following paragraph.

> The ``public/uploads`` directory can now be read from and write to as usual and

Reviewer 3: `can now be read ... and write ..` -> `can now be read ... and written ..` ?

> several containers can acces this dictory and share the **same** files. The

Reviewer 2: directory

> several containers can acces this dictory and share the **same** files. The

Reviewer 3: `acces` -> `access`

> ``var`` directory however stays "local" and will not be shared by the different

Reviewer 3: `shared by` -> `shared among` ?  (or `shared between` ?)

   You do **not** need to add a relationship to point to the ``files`` service.
   This is handled automatically by SymfonyCloud.

The application can now read from and write to the ``public/uploads`` directory

> as if it was a local writeable mount.

Reviewer 2: I think we need to clearly state the difference between `var` and `/pub/uploads` in your example.

Author: updated. let me know if it is better now

.. note::

    There is a small performance impact for using a network mount over a local
    mount, but in most cases it should not be noticeable. Please note that
    high-volume sequential file creation may experience a more significant
    performance impact.

How can I migrate a Local Storage to a Network Storage?
-------------------------------------------------------

There is no automated way of transferring data from one storage type to

> another. Fundamentally, the process is consists in moving files around.

Reviewer 2: I would remove the 'Fundamentally' sentence.

Suppose you have this mount configuration and want to move to a network storage
mount:

.. code-block:: yaml

    mounts:
        '/var': { source: local, source_path: var }
        '/public/uploads': { source: local, source_path: uploads }

> The following approximate steps will do so with a minimum of service interruption.

Reviewer 2: I would remove "approximate"

The following steps will do so with a minimum of service interruption.

> 1. Add a new ``network-storage`` service named ``files``, that has at least

Reviewer: ```suggestion
#. Add a new ``network-storage`` service named ``files``, that has at least
```

   enough space for your existing files with some buffer.

   .. caution::

       You may need to increase your project's allocated storage to do this.
       This operation is not reversible and will incur charges.

> 2. Add a new mount to the network storage service on a non-public directory:

Reviewer: ```suggestion
#. Add a new mount to the network storage service on a non-public directory:
```

   .. code-block:: diff

       mounts:
           '/var': { source: local, source_path: var }
           '/public/uploads': { source: local, source_path: uploads }
    +      '/new-uploads': { source: service, service: files, source_path: uploads }

> 3. Deploy the changes;

Reviewer: ```suggestion
#. Deploy the changes;
```

> 4. Connect to your container using ``symfony ssh``;

Reviewer: ```suggestion
#. Connect to your container using ``symfony ssh``;
```

> 5. Use ``rsync`` to copy all files from the local mount to the network mount:

Reviewer: ```suggestion
#. Use ``rsync`` to copy all files from the local mount to the network mount:
```

   .. code-block:: terminal

      $ rsync -avz web/uploads/* new-uploads/

> 6. Reverse the mounts, commit and deploy:

Reviewer: ```suggestion
#. Reverse the mounts, commit and deploy:
```

   .. code-block:: diff

       mounts:
           '/var': { source: local, source_path: var }
    -      '/public/uploads': { source: local, source_path: uploads }
    +      '/old-uploads': { source: local, source_path: uploads }
    -      '/new-uploads': { source: service, service: files, source_path: uploads }

>     +      '/public/uploads': { source: service, service: files, source_path: uploads }

Author: I know this is not strictly a compatible diff syntax but I think it makes more sense visually.
WDYT?

Reviewer: Looks good to me 👍

    +      '/public/uploads': { source: service, service: files, source_path: uploads }

> 7. Run ``rsync`` again to make sure all files uploaded during the transition are moved.

Reviewer: ```suggestion
#. Run ``rsync`` again to make sure all files uploaded during the transition are moved.
```

   .. code-block:: terminal

      $ rsync -avz old-uploads/* web/uploads/

8. Once you're confident that all files are accounted for, delete the entire

>    contents of ``old-uploads``. If you do not, the files will remain on disk

Reviewer 2: I woudl remove the last sentence.

> 8. Once you're confident that all files are accounted for, delete the entire

Reviewer: ```suggestion
#. Once you're confident that all files are accounted for, delete the entire
```

   contents of ``old-uploads``.

> 9. Once done, you can remove the ``old-uploads`` mount and push again to

Reviewer: ```suggestion
#. Once done, you can remove the ``old-uploads`` mount and push again to
```
